### PR TITLE
Adds automatic chunking and (basic) output content filtering API

### LINKED ISSUE #1327: Add support for chunked responses on test server
I believe it would be a good addition to werkzeug's internal testing server if it were able to process chunked HTTP responses.
I already implemented the protocol not long ago and I believe I could contribute with such code.

### PR DESCRIPTION
**Pull request by brunoais**
Here I have a proposal for automatic chunking detection with nearly the lowest amount of code required.

This PR is incomplete on purpose (see below). As it is waiting for feedback from the developers.
Todo:
- [x] Make minimal testable content for review
- [x] Confirm interest in the feature from the maintainers
- [ ] Create a "kill-switch" option to revert to previous behavior (probably in environ)
- [ ] Make automatic tests

closes #1327

### CODE REVIEW

File: werkzeug/serving.py [hunk 1]
```diff
@@ -246,8 +258,16 @@ def write(data):
                     self.send_header('Date', self.date_time_string())
                 self.end_headers()
 
+                if 'transfer-encoding' in header_keys:
+                    for key, value in response_headers:
+                        if 'transfer-encoding' == key.lower():
```
Comment (ThiefMaster) [inline werkzeug/serving.py]: please don't do yoda style. it's awful.

Comment (brunoais) [reply werkzeug/serving.py]: @ThiefMaster Do you mean to place the string on the right?

Comment (ThiefMaster) [reply werkzeug/serving.py]: Yes, `'constant' == variable` is yoda style

**Comment by brunoais** [comment]
The tests currently not passing.... Is it caused by my code?

**Comment by brunoais** [comment]
@ThiefMaster Is there interest in this PR?

File: werkzeug/serving.py [hunk 2]
```diff
@@ -219,6 +219,19 @@ def run_wsgi(self):
         self.environ = environ = self.make_environ()
         headers_set = []
         headers_sent = []
+        filters_applied = set()
+
+        # Replaces nonlocal for python 2.7
+        class nonlocal_(object):
```
Comment (ThiefMaster) [inline werkzeug/serving.py]: that seems overly complicated, especially with the staticmethod() below. why not just use `nonlocal_ = [lambda data: data]` here and then `nonlocal_[0]` to access/reassign?

**Comment by ThiefMaster** [comment]
Sure looks like a useful feature, but I'd rather have someone more familiar with chunking and when it's required/useful look at this.

File: werkzeug/serving.py (hunk 2 above)
Comment (brunoais) [reply werkzeug/serving.py]: Sure. Will do in the next commit

Comment (brunoais) [reply werkzeug/serving.py]: @ThiefMaster pushed the change as you suggested.

**Comment by brunoais** [comment]
Great, then!
When able, I'll build a "kill-switch" which allows people to opt-out of this feature just in case. Regardless, this code is only for the werkzeug development server... Even so, if it gets accepted it will help me a lot to develop a few other servers which expect chunked responses

**Comment by brunoais** [comment]
Update: Got issues in life that slowed this down. Thankfully, I know I will have time to work on this again starting on 2nd week of September. I hope to get the killswitch thought so someone may review this PR.

**Comment by davidism** [comment]
Thanks for working on this. However, I'm deciding not to merge it.

PEP 3333 specifically says that the application must not set hop-by-hop headers such as `Transfer-Encoding`. https://www.python.org/dev/peps/pep-3333/#other-http-features If the server were to support it, it would decide to based on the `Content-Length` header, and set the `Transfer-Encoding` header itself.

Then, it comes down to whether we want the dev server to support sending chunked encoding. As far as I can tell this just affects how connections are kept alive. Sending a stream as supported now just closes the connection afterwards. Given that to the end user the received data looks the same, I don't really see the benefit of sending chunked encoding from the dev server.

Also, this introduces more complexity with the filter logic, which doesn't seem necessary to the feature or make clear what other features might be require this.

Dedicated WSGI servers are better suited to handle this. If you'd like to have your streamed response returned as a chunked encoding, something like Gunicorn will be able to handle that. That can still be used during development, just not with `run_simple`.

**Pull request closed without merging**